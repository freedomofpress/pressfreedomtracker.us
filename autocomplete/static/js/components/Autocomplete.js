import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import axios from 'axios'

import Single from './Single'
import Multi from './Multi'


class Autocomplete extends PureComponent {
	constructor(props, ...args) {
		super(props, ...args)

		this.handleClick = this.handleClick.bind(this)
		this.handleChange = this.handleChange.bind(this)
		this.handleCreate = this.handleCreate.bind(this)

		this.state = {
			value: props.value,
			input: {
				value: '',
			},
			suggestions: [],
		}

		if (props.fetchInitialValues) {
			this.fetchInitialValues(props.value)
		}
	}

	componentDidMount() {
		this.checkNewSuggestions('', false)
	}

	get value() {
		if (this.props.controlled) {
			return this.props.value
		} else {
			return this.state.value
		}
	}

	handleChange(event) {
		const { value } = event.target
		this.checkNewSuggestions(value)
		this.setState({
			input: {
				...this.state.input,
				value,
			}
		})
	}

	getExclusions() {
		const { value } = this.state
		if (!value) {
			return ''
		}

		if (this.props.isSingle && value) {
			return value.id
		}

		return value.map(({ id }) => id).join(',')
	}

	checkNewSuggestions(value, checkDifferent = true) {
		if (checkDifferent && value === this.state.value) {
			return
		}

		const params = {
			query: value,
			type: this.props.type,
			exclude: this.getExclusions(),
		}
		axios.get(this.props.apiBase + 'search/', { params })
			.then(res => {
				if (res.status !== 200) {
					return
				}

				if (!Array.isArray(res.data.pages)) {
					return
				}

				this.setState({
					suggestions: res.data.pages
				})
			})
	}

	fetchInitialValues(value) {
		if (!value) {
			return
		}

		const isMulti = Array.isArray(value)
		if (isMulti && value.length === 0) {
			return
		}

		if (isMulti) {
			var ids = value.map(({ id }) => id).join(',')
		} else {
			var ids = value.id
		}

		const params = {
			ids,
			type: this.props.type,
		}
		axios.get(this.props.apiBase + 'objects/', { params })
			.then(res => {
				if (res.status !== 200) {
					return
				}

				if (!Array.isArray(res.data.pages)) {
					return
				}

				if (isMulti) {
					var value = this.state.value.map(value => {
						const page = res.data.pages.find(page => page.id === value.id)
						if (!page) {
							return value
						}

						return page
					})
				} else {
					var value = res.data.pages[0]
				}

				this.setState({ value })

				if (typeof this.props.onChange === 'function') {
					this.props.onChange({ target: { value } })
				}
			})
	}

	handleClick(value) {
		this.setState({ value })

		if (typeof this.props.onChange === 'function') {
			this.props.onChange({ target: { value, _autocomplete: true } })
		}
	}

	handleCreate() {
		const { value } = this.state.input
		if (value.trim() === '') {
			return
		}

		const data = new FormData()
		data.set('type', this.props.type)
		data.set('value', value)
		axios.post(this.props.apiBase + 'create/', data)
			.then(res => {
				if (res.status !== 200) {
					this.setState({ isLoading: false })
					return
				}

				const value = this.props.isSingle ? res.data : (this.state.value || []).concat(res.data)

				this.setState({
					isLoading: false,
					value,
				})

				if (typeof this.props.onChange === 'function') {
					this.props.onChange({ target: { value } })
				}
			})
		this.setState({ isLoading: true })
	}

	render() {
		const { name, isSingle, onChange } = this.props
		const { input, suggestions } = this.state

		const canCreate = this.props.canCreate && input.value.trim() !== ''
		const useHiddenInput = typeof onChange !== 'function'

		return (
			<span className="autocomplete">
				{useHiddenInput && (
					<input
						type="hidden"
						value={JSON.stringify(this.value)}
						name={name}
					/>
				)}

				{isSingle && (
					<Single
						input={input}
						suggestions={suggestions}
						selected={this.value}

						canCreate={canCreate}

						onCreate={this.handleCreate}
						onChange={this.handleChange}
						onClick={this.handleClick}
					/>
				)}

				{!isSingle && (
					<Multi
						input={input}
						suggestions={suggestions}
						selections={this.value || Multi.defaultProps.selections}

						canCreate={canCreate}

						onCreate={this.handleCreate}
						onChange={this.handleChange}
						onClick={this.handleClick}
					/>
				)}
			</span>
		)
	}
}


Autocomplete.defaultProps = {
	fetchInitialValues: false,
	controlled: false,
}


Autocomplete.propTypes = {
	name: PropTypes.string.isRequired,
	type: PropTypes.string.isRequired,
	canCreate: PropTypes.bool.isRequired,
	isSingle: PropTypes.bool.isRequired,
	onChange: PropTypes.func,
	fetchInitialValues: PropTypes.bool,
	apiBase: PropTypes.string.isRequired,
	controlled: PropTypes.bool.isRequired,
}


export default Autocomplete
