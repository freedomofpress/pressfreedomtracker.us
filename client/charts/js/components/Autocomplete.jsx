import React, {useReducer, useEffect} from 'react'
import classNames from "classnames"

import PropTypes from 'prop-types'


const initialState = {
	// The active selection's index
    activeSuggestion: 0,
    // The suggestions that match the user's input
    filteredSuggestions: [],
    // Whether or not the suggestion list is shown
    showSuggestions: false,
    // What the user has entered
    userInput: ""
}

const propTypes = {
	suggestions: PropTypes.array.isRequired,
	suggestionsLabelField: PropTypes.string.isRequired,
	suggestionsSidenoteField: PropTypes.string.isRequired,
	name: PropTypes.string.isRequired,
	placeholder: PropTypes.string.isRequired,
	handleSelect: PropTypes.func.isRequired,
	itemNamePlural: PropTypes.string.isRequired,
}

const filterSuggestions = query => ({
	type: 'FILTER_SUGGESTIONS',
	payload: query,
})


const selectSuggestion = query => ({
	type: 'SELECT_SUGGESTION',
	payload: query,
})


const incrementActiveSuggestion = {
	type: 'INCREMENT_ACTIVE_SUGGESTION',
	payload: null,
}


const decrementActiveSuggestion = {
	type: 'DECREMENT_ACTIVE_SUGGESTION',
	payload: null,
}


const AutoComplete = ({
	suggestions,
	suggestionsLabelField,
	suggestionsSidenoteField,
	name,
	placeholder,
	handleSelect,
	itemNameSingular,
	itemNamePlural,
}) => {
	const reducer = (state, {type, payload}) => {
		switch (type) {
			case 'SELECT_SUGGESTION':
				return initialState
			case 'DECREMENT_ACTIVE_SUGGESTION':
				return {...state, activeSuggestion: Math.max(0, state.activeSuggestion - 1)}
			case 'INCREMENT_ACTIVE_SUGGESTION':
				return {...state, activeSuggestion: Math.min(state.filteredSuggestions.length - 1, state.activeSuggestion + 1)}
			case 'FILTER_SUGGESTIONS':
				const filteredSuggestions = suggestions.filter(
					suggestion =>
					suggestion[suggestionsLabelField].toLowerCase().indexOf(payload.toLowerCase()) > -1
				)
				return {
					activeSuggestion: 0,
					filteredSuggestions,
					userInput: payload,
					showSuggestions: true,
				}

			default:
				throw new Error(`Unknown action type: ${type}`)
		}
	}

	const [state, dispatch] = useReducer(reducer, initialState)

	const id = `id_${name}`
	const choicesId = `${name}_choices`
	const onChange = (event) => dispatch(filterSuggestions(event.target.value))
	const onClick = (event) => {
		let val = event.currentTarget.dataset.item
		handleSelect(val)
		dispatch(selectSuggestion(val))
	}

	const onKeyDown = (event) => {
		if (event.keyCode === 13) { // Enter key
			let val = state.filteredSuggestions[state.activeSuggestion][suggestionsLabelField]
			handleSelect(val)
			dispatch(selectSuggestion(val))

		} else if (event.keyCode === 38) { // up arrow
			dispatch(decrementActiveSuggestion)
		} else if (event.keyCode === 40) { // down arrow
			dispatch(incrementActiveSuggestion)
		}
	}

	let suggestionListComponent
	if (state.showSuggestions && state.userInput) {
		if (state.filteredSuggestions.length) {
			suggestionListComponent = (
				<ul className="filters__suggestions">
					{state.filteredSuggestions.map((suggestion, index) => (
						<li
							className="filters__suggestions-item filters__suggestions-item--selectable"
							className={classNames("filters__suggestions-item", "filters__suggestions-item--selectable", {"filters__suggestions-item--active": index === state.activeSuggestion})}
							onClick={onClick}
							data-item={suggestion[suggestionsLabelField]}
							key={index}
						>
							<span>
								{suggestion[suggestionsLabelField]}
								<span
									className="filters__suggestion-tooltip"
								>
									&ndash;
									<span
										className="filters__suggestion-tooltip-text"
									>
										Add {itemNameSingular}
									</span>
								</span>
							</span>
							<span>{suggestion[suggestionsSidenoteField]}</span>
						</li>
					))}
				</ul>
			)
		} else {
			suggestionListComponent = (
				<ul className="filters__suggestions">
					<li className="filters__suggestions-item">
						No {itemNamePlural} found
					</li>
				</ul>
			)
		}
	}

	return (
		<>
			<input
				type="text"
				name={name}
				className="text-field--single"
				autoComplete="off"
				onChange={onChange}
				onKeyDown={onKeyDown}
				id={id}
				placeholder={placeholder}
				value={state.userInput}
			/>
			{suggestionListComponent}
		</>
	)
}


AutoComplete.propTypes = propTypes

export default AutoComplete
