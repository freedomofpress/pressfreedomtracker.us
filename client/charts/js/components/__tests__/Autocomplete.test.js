import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'
import Autocomplete from '../Autocomplete'

test('renders Autocomplete with mocked data', () => {
	const renderer = new ShallowRenderer();
	expect(renderer.render(
		<Autocomplete
			suggestions={[{label: "test label"}]}
			suggestionsLabelField={'label'}
			placeholder={'test'}
			name={'test autocomplete'}
			itemNamePlural={'tests'}
			handleSelect={() => {}}
			suggestionsSidenoteField={'sidenote'}
		/>
	)).toMatchSnapshot();
});
