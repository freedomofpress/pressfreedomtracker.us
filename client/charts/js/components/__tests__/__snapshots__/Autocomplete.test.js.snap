// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Autocomplete with mocked data 1`] = `
<React.Fragment>
  <input
    autoComplete="off"
    className="text-field--single"
    id="id_test autocomplete"
    name="test autocomplete"
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="test"
    type="text"
    value=""
  />
</React.Fragment>
`;
